### ISSUE TITLE
[explicit-function-return-type] Only apply to generic functions

### ISSUE THREAD

**Issue by AnyhowStep** [typescript-eslint/typescript-eslint#730]
The template seems to be for bug reports but I have a feature request.

I would like a way to configure `explicit-function-return-type` so that it only forces generic functions to have an explicit return type.

```ts
//Error, generic functions must have explicit return type
function foo<T extends SomeType> (t : T) {
    return /*snip some code using `t`*/;
}

//OK! Not generic, so no need for explicit return type
function blah (t : SomeType) {
    return /*snip some code using `t`*/;
}
```

### Rationale

In general, `tsc`'s type inference is very fast.

However, when it comes to complicated generic types, type inference can fail and/or infer... unfavorable types. This is especially true of generic function return types.

I personally just had a check time performance issue that was fixed by adding an explicit return type annotation to one generic function.

It cut down check time by **five minutes**

-----

This missing one line of code caused me weeks of suffering =(

I do not wish for others and future me to trip over this again =(

In general, I tend to always add explicit retun types to all my generic functions. The **one** time I forgot, it messed me up =(

-----

### Letting `tsc` infer return type

![image](https://user-images.githubusercontent.com/5655961/61577512-63676100-aab6-11e9-884e-6f0e5ea088a9.png)

### Explicitly setting return type

![image](https://user-images.githubusercontent.com/5655961/61577518-724e1380-aab6-11e9-974c-9c833dd482f2.png)

The difference is staggering

**Comment by bradzacher** [comment]
man what code are you writing that you have such complex, messed up types!?!?!?

**Comment by AnyhowStep** [comment]
I've been experimenting with different ways to squeeze out more safety from my code. 

Sometimes, it leads to huge performance trade offs :x

Other times, I just forget to add return type annotations.

In this case, I was specifically working on a project that uses this library.

https://github.com/anyhowstep/type-mapping

It's kind of like io-ts, but worse in some areas and better in others, if I had to describe it

**Comment by patrickmichalina** [comment]
Is this being considered? Really sucks to have to explicitly type all function returns to avoid this memory issue with the `incremental` flag.

**Comment by bradzacher** [comment]
this project is community maintained - that means that the things that get built are the features the community finds value in.
we maintainers are volunteers (and very few), so we focus our time on doing "big ticket" tasks.

happy to accept a PR that adds an option to the rule like `onlyCheckFunctionsWithTypeParameters`, which causes the rule to only match functions with generic a type parameter declaration.

**Comment by patrickmichalina** [comment]
@bradzacher any existing rules that could give someone a starting point if they want to tackle this issue?

**Comment by bradzacher** [comment]
I'd spend some time poking around in our rules to get the gist of it.

http://astexplorer.net is your best friend here for figuring out the structure of the ast so you can understand how you might limit the scope.
